cookbooks/bcpc-hadoop/libraries/utils.rb

```diff
@@ -203,59 +203,84 @@ def set_hosts
+    nodes.select { |n| runs_role.call(n, 'job_history_server') }.map { |n| to_host.call(n) }
 
   node.default[:bcpc][:hadoop][:dn_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hadoop_worker'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hadoop_worker') }.map { |n| to_host.call(n) }
 
   node.default[:bcpc][:hadoop][:hb_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hbase_head'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hbase_head') }.map { |n| to_host.call(n) }
 
   # different flavors of hive
   node.default[:bcpc][:hadoop][:hive_hosts] = 
-      cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hive_server'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hive_server') }.map { |n| to_host.call(n) }
 
-  # BCPC-Hadoop-Head-MapReduce
+  # (the BCPC-Hadoop-Head-MapReduce role nests the oozie recipe)
   node.default[:bcpc][:hadoop][:oozie_hosts]  = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['oozie_server'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
-  
+    nodes.select do |n|
+      runs_role.call(n, 'oozie_server') || runs_recipe.call(n, 'oozie_server')
+    end.map { |n| to_host.call(n) }
+
   # every datanode
   node.default[:bcpc][:hadoop][:httpfs_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hadoop_worker']  }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hadoop_worker') }.map { |n| to_host.call(n) }
 
-  # Worker
+  # every worker
   node.default[:bcpc][:hadoop][:rs_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hadoop_worker']  }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hadoop_worker') }.map { |n| to_host.call(n) }
 
   # BCPC-Hadoop-Head
   node.default[:bcpc][:hadoop][:mysql_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hadoop_head'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hadoop_head') }.map { |n| to_host.call(n) }
+
+  # set the oozie_url
+  oozie_hosts = node[:bcpc][:hadoop][:oozie_hosts]
+  if oozie_hosts.length > 1
+    # high-availability
+    node.default[:bcpc][:hadoop][:oozie_url] =
+      "http://#{float_host(node[:bcpc][:management][:viphost])}:" +
+        "#{node[:bcpc][:ha_oozie][:port]}/oozie"
+  elsif oozie_hosts.length == 1
+    # single oozie host
+    node.default[:bcpc][:hadoop][:oozie_url] =
+      "http://#{float_host(oozie_hosts.first[:hostname])}:" +
+        "#{node[:bcpc][:hadoop][:oozie_port]}/oozie"
+  end
+
+  # set the HA resourcemanager_url (rm_address)
+  rm_hosts = node[:bcpc][:hadoop][:rm_hosts]
+  node.default[:bcpc][:hadoop][:rm_address] = node.chef_environment
```

Comment:
Should not this check that we have more than one RM?

Reply:
That's how I originally had it, but this should work if we want to throw errors when there's no HA RM. 
I can change it back to work with a single RM too. @aespinosa thoughts?